src/plugins/data/common/search/aggs/buckets/lib/histogram_calculate_interval.ts

```diff
@@ -0,0 +1,142 @@
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { isAutoInterval } from '../_interval_options';
+
+interface IntervalValuesRange {
+  min: number;
+  max: number;
+}
+
+export interface CalculateHistogramIntervalParams {
+  interval: string;
+  maxBucketsUiSettings: number;
+  maxBucketsUserInput?: number;
+  intervalBase?: number;
+  values?: IntervalValuesRange;
+}
+
+/**
+ * Round interval by order of magnitude to provide clean intervals
+ */
+const roundInterval = (minInterval: number) => {
+  const orderOfMagnitude = Math.pow(10, Math.floor(Math.log10(minInterval)));
+  let interval = orderOfMagnitude;
+
+  while (interval < minInterval) {
+    interval += orderOfMagnitude;
+  }
+
+  return interval;
+};
+
+const calculateForGivenInterval = (
+  diff: number,
+  interval: number,
+  maxBucketsUiSettings: CalculateHistogramIntervalParams['maxBucketsUiSettings']
+) => {
+  const bars = diff / interval;
+
+  if (bars > maxBucketsUiSettings) {
+    const minInterval = diff / maxBucketsUiSettings;
+
+    return roundInterval(minInterval);
+  }
+
+  return interval;
+};
+
+const calculateAutoInterval = (
+  diff: number,
+  values: IntervalValuesRange,
+  maxBucketsUiSettings: CalculateHistogramIntervalParams['maxBucketsUiSettings'],
+  maxBucketsUserInput: CalculateHistogramIntervalParams['maxBucketsUserInput']
+) => {
+  const getMaxFractionalPart = (...numbers: number[]) => {
+    return numbers.reduce((acc, n) => {
+      const [, value] = n.toString().split('.');
+      return value && value.length > acc ? value.length : acc;
+    }, 0);
+  };
+
+  let autoBuckets: number = 0;
+
+  if (!maxBucketsUserInput) {
+    const fractalFactor = Math.pow(10, getMaxFractionalPart(values.min, values.max));
+
+    autoBuckets = diff * fractalFactor;
+
+    if (autoBuckets > maxBucketsUiSettings) {
+      autoBuckets =
+        (Math.log10(maxBucketsUiSettings) / Math.log10(autoBuckets)) * maxBucketsUiSettings;
+    }
+
+    // For float numbers we can try to increase interval
+    if (fractalFactor > 1 && autoBuckets < maxBucketsUiSettings / 2) {
+      autoBuckets = autoBuckets * 5;
+      if (autoBuckets > maxBucketsUiSettings) {
+        autoBuckets = autoBuckets * 2;
+      }
+    }
+  }
+
+  const bars = Math.min(maxBucketsUiSettings, maxBucketsUserInput || autoBuckets || 1);
+
+  return roundInterval(diff / bars);
+};
+
+export const calculateHistogramInterval = ({
```

Comment:
This is the type of code that can be really hard to maintain if you have to read through the math without any background and grok what's going on. 

For future reference & to make maintenance easier, it would be great if we could add some detailed comments to each of the functions in this file, explaining the algorithm at a high level, what each function does, and why.

Reply:
Good idea, I've  added a textual description of @wylieconlon 's original algorithm https://github.com/elastic/kibana/pull/76001/commits/11f6ec8ea328b3ecab2b6713fc5e8a46b8f84560